x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -143,12 +150,106 @@ public void execute(
+        private final EsqlQueryRequest request;
+        private final EsqlExecutionInfo executionInfo;
+        private final BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase;
+        private final ActionListener<Result> listener;
+
+        LogicalPlanActionListener(
+            EsqlQueryRequest request,
+            EsqlExecutionInfo executionInfo,
+            BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
+            ActionListener<Result> listener
+        ) {
+            this.request = request;
+            this.executionInfo = executionInfo;
+            this.runPhase = runPhase;
+            this.listener = listener;
+        }
+
+        @Override
+        public void onResponse(LogicalPlan analyzedPlan) {
+            executeOptimizedPlan(request, executionInfo, runPhase, optimizedPlan(analyzedPlan), listener);
+        }
+
+        /**
+         * Whether to return an empty result (HTTP status 200) for a CCS rather than a top level 4xx/5xx error.
+         *
+         * For cases where field-caps had no indices to search and the remotes were unavailable, we
+         * return an empty successful response (200) if all remotes are marked with skip_unavailable=true.
+         *
+         * Note: a follow-on PR will expand this logic to handle cases where no indices could be found to match
+         * on any of the requested clusters.
+         */
+        private boolean returnSuccessWithEmptyResult(Exception e) {
+            if (executionInfo.isCrossClusterSearch() == false) {
+                return false;
+            }
+
+            if ((e instanceof IllegalStateException ill && ill.getMessage().equals("No clusters to search"))
+                || ExceptionsHelper.isRemoteUnavailableException(e)) {
+                for (String clusterAlias : executionInfo.clusterAliases()) {
+                    if (executionInfo.isSkipUnavailable(clusterAlias) == false
+                        && clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY) == false) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (returnSuccessWithEmptyResult(e)) {
+                executionInfo.markEndQuery();
+                Exception exceptionForResponse;
+                if (e instanceof ConnectTransportException) {
+                    // when field-caps has no field info (since no clusters could be connected to or had matching indices)
+                    // it just throws the first exception in its list, so this odd special handling is here is to avoid
+                    // having one specific remote alias name in all failure lists in the metadata response
+                    exceptionForResponse = new RemoteTransportException(
+                        "connect_transport_exception - unable to connect to remote cluster",
```

Comment:
If you have two clusters, both skip_un=true, and field-caps cannot connect to either of them, rather than returning both errors, field-caps returns one (arbitrarily). If you just put that error message in the ExecutionInfo.Cluster, you get this incorrect and confusing output:

```
 "_clusters": {
    "total": 2,
    "successful": 0,
    "running": 0,
    "skipped": 2,
    "partial": 0,
    "failed": 0,
    "details": {
      "remote2": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "connect_transport_exception",
              "reason": "[could not connect to remote1]"
            }
          }
        ]
      },
      "remote1": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "connect_transport_exception",
              "reason": "[could not connect to remote1]"
            }
          }
        ]
      }
    }
```

This special handling changes it to be a generic "could not connect to remote cluster" message, like so:

```
 "_clusters": {
    "total": 2,
    "successful": 0,
    "running": 0,
    "skipped": 2,
    "partial": 0,
    "failed": 0,
    "details": {
      "remote2": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      },
      "remote1": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      }
    }
```

Reply:
> This message is a bit confusing to me - is it about connecting to some specific cluster
The error message occurs within the cluster/details of a specific cluster, so the intention is that it is obvious which cluster we could not connect to.

I could add additional handling to swap in the name of each cluster into the error message if you think this model is not clear.

Reply:
No if it's inside specific cluster's data then it's fine I think.